keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+the policy is initially set to the most restrictive state and can be made less
+restrictive via configuration.
+
+## Proposal
+
+Introduce a new `ValidatingAdmissionPolicy` kind to the
+admissionregistration.k8s.io group. (suggestions welcome on exact name to use
+for kind)
+
+At a high level, the API will support:
+
+- Request matching (similar to webhook Rules, NamespaceSelector,
+  ObjectSelector)
+- CEL rule evaluation (similar to both [CRD Validation
+  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+  and
+  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+- Automatic version conversion support (similar to webhooks MatchPolicy=Equivalent), with ability to introspect GVK of admitted resource in CEL expression.
+- Access the old object (similar to [transition
+  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
+  and oldObject in AdmissionRequest)
+- Configurability, as motivated above.
+
+There are also lots of additional capabilities (response message formatting,
+severity, levels, failure policies, type safety, ...) that will be discussed
+in detail further on in this proposal.
+
+We have divided this proposal into phases, all of which must be completed before
+this feature graduates to beta. Our goal is to size the phases so that each
+can be completed in a single Kubernetes release cycle.
+
+### Phase 1
+
+#### API Shape
+
+Before getting into all the individual fields and capabilities, let's look at the
+general "shape" of the API.
+
+This enhancement introduces a new `ValidatingAdmissionPolicy` kind.
+
+Each `ValidatingAdmissionPolicy` resource defines a admission control policy.
+The resource contains the CEL expressions to validate the admission policy and
+declares how the admission policy may be configured for use.
+
+For example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  config:
+    group: rules.example.com
+    kind: ReplicaLimit
+    version: v1
+  constraints:
+    rules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["deployments"]
+      scope:       "*"
+  validations:
+    - expression: "object.spec.replicas <= config.maxReplicas"
+      # ...other rule related fields here...
+```
+
+The `spec.config` field of the `ValidatingAdmissionPolicy` references the
+"configuration CRD" for this admission policy. For this example, the
+configuration CRD is `ReplicaLimit`.
+
+Note: This is a "Bring Your Own CRD" design. The admission policy definition
+author is responsible for providing the `ReplicaLimit` configuration CRD.
```

Comment:
Will there be any automatic admission on CRDs to ensure the schemas when modified conform to the expectations of the ValidatingAdmissionPolicy?

Reply:
I wasn't planning on using admission control.  I want to support CRDs and ValidatingAdmissionPolicies being created in any order.  If the ValidatingAdmissionPolicy is created first, it will need to exist in a "misconfigured" state until the CRD is created.